File: src/web/magmac/api/None.ts
export class None<T> {
	public map( mapper : Function<T, R>) : Option<R> {return new None<>( );;}
	public isPresent() : boolean {return false;;}
	public orElseGet( other : Supplier<T>) : T {return other.get( );;}
	public isEmpty() : boolean {return true;;}
	public flatMap( mapper : Function<T, Option<R>>) : Option<R> {return new None<>( );;}
	public orElse( other : T) : T {return other;;}
	public filter( predicate : Predicate<T>) : Option<T> {return this;;}
	public or( other : Supplier<Option<T>>) : Option<T> {return other.get( );;}
	public ifPresent( consumer : Consumer<T>) : void {;}
}

File: packages/compiler-java/src/main/java/magmac/api/None.java
package magmac.api;

import java.util.function.Consumer;
import java.util.function.Function;
import java.util.function.Predicate;
import java.util.function.Supplier;

/**
 * Option variant representing the absence of a value.
 */

public class None<T> implements Option<T> {
    @Override
    public <R> Option<R> map(Function<T, R> mapper) {
        return new None<>();
    }

    @Override
    public boolean isPresent() {
        return false;
    }

    @Override
    public T orElseGet(Supplier<T> other) {
        return other.get();
    }

    @Override
    public boolean isEmpty() {
        return true;
    }

    @Override
    public <R> Option<R> flatMap(Function<T, Option<R>> mapper) {
        return new None<>();
    }

    @Override
    public T orElse(T other) {
        return other;
    }

    @Override
    public Option<T> filter(Predicate<T> predicate) {
        return this;
    }

    @Override
    public Option<T> or(Supplier<Option<T>> other) {
        return other.get();
    }

    @Override
    public void ifPresent(Consumer<T> consumer) {
    }
}
